src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/ReferenceAssemblies.cs

```diff
@@ -123,14 +125,83 @@ public static ReferenceAssemblies Default
+        {
+            if (s_knownAssemblies.TryGetValue(value, out var existingValue))
+            {
+                return existingValue;
+            }
+
+            if (ImmutableInterlocked.Update(
+                ref s_knownAssemblies,
+                static (knownAssemblies, value) => knownAssemblies.Add(value),
+                value))
+            {
+                return value;
+            }
+
+            if (!s_knownAssemblies.TryGetValue(value, out existingValue))
+            {
+                throw new InvalidOperationException();
+            }
+
+            return existingValue;
+        }
+
+        public override int GetHashCode()
+        {
+#if NETCOREAPP
+            var hash = default(HashCode);
+            hash.Add(TargetFramework);
+            hash.Add(AssemblyIdentityComparer);
+            hash.Add(ReferenceAssemblyPackage);
+            hash.Add(ReferenceAssemblyPath);
+            hash.Add(Assemblies, ImmutableArrayEqualityComparer<string>.Instance);
+            hash.Add(FacadeAssemblies, ImmutableArrayEqualityComparer<string>.Instance);
+            hash.Add(LanguageSpecificAssemblies, ImmutableDictionaryWithImmutableArrayValuesEqualityComparer<string, string>.Instance);
+            hash.Add(Packages, ImmutableArrayEqualityComparer<PackageIdentity>.Instance);
+            hash.Add(NuGetConfigFilePath);
+            return hash.ToHashCode();
+#else
+            var hashCode = -450793227;
+            hashCode = (hashCode * -1521134295) + EqualityComparer<string>.Default.GetHashCode(TargetFramework);
+            hashCode = (hashCode * -1521134295) + EqualityComparer<AssemblyIdentityComparer>.Default.GetHashCode(AssemblyIdentityComparer);
+            hashCode = (hashCode * -1521134295) + EqualityComparer<PackageIdentity?>.Default.GetHashCode(ReferenceAssemblyPackage);
+            hashCode = (hashCode * -1521134295) + EqualityComparer<string?>.Default.GetHashCode(ReferenceAssemblyPath);
+            hashCode = (hashCode * -1521134295) + ImmutableArrayEqualityComparer<string>.Instance.GetHashCode(Assemblies);
+            hashCode = (hashCode * -1521134295) + ImmutableArrayEqualityComparer<string>.Instance.GetHashCode(FacadeAssemblies);
+            hashCode = (hashCode * -1521134295) + ImmutableDictionaryWithImmutableArrayValuesEqualityComparer<string, string>.Instance.GetHashCode(LanguageSpecificAssemblies);
+            hashCode = (hashCode * -1521134295) + ImmutableArrayEqualityComparer<PackageIdentity>.Instance.GetHashCode(Packages);
+            hashCode = (hashCode * -1521134295) + EqualityComparer<string?>.Default.GetHashCode(NuGetConfigFilePath);
+            return hashCode;
+#endif
+        }
+
+        public override bool Equals(object? obj)
+            => Equals(obj as ReferenceAssemblies);
+
+        public bool Equals(ReferenceAssemblies? other)
+        {
+            return other is not null
```

Comment:
>return other is not null [](http://example.com/codeflow?start=12&length=24)

maybe a reference equality check preceeding all the other work?

Reply:
oh, it looks like the GetOrAddReferenceAssemblies call takes in a newly constructed one, so wouldn't really have any benefit